### PR TITLE
Add PytorchProfilerCallback

### PR DESCRIPTION
**Pull request by TaekyungHeo**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occasions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do?
This PR introduces the `ChakraCallback`, a new PyTorch Lightning callback that collects Chakra host and device traces during training.

**Collection**: `nemo.lightning.pytorch.callbacks`

# Changelog
- Added `ChakraCallback` to `nemo.lightning.pytorch.callbacks`.
- Implemented profiling support for capturing PyTorch execution traces (host) and Kineto traces (device).
- Updated `__init__.py` to include `ChakraCallback`.

# Usage

Example of how to use the `ChakraCallback`:

```python
from nemo.lightning.pytorch.callbacks.chakra import ChakraCallback

chakra_callback = ChakraCallback(
    start_step=5,
    end_step=10,
    warmup_steps=1,
    active_steps=1,
    trace_dir="/workspace/"
)

trainer = Trainer(callbacks=[chakra_callback])
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

# Additional Information
* Related PR: https://github.com/NVIDIA/NeMo/pull/11115

### CODE REVIEW

File: nemo/lightning/pytorch/callbacks/chakra.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+from pathlib import Path
+
+import torch
+from lightning.pytorch.callbacks import Callback
+
+from nemo.utils import logging
+from nemo.utils.get_rank import get_rank
+
+
+class ChakraCallback(Callback):
```
Comment (titu1994) [inline nemo/lightning/pytorch/callbacks/chakra.py:25]: Rename to PytorchProfilerCallback rather than some code name.

File: nemo/lightning/pytorch/callbacks/chakra.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        self.trace_dir = Path(trace_dir)
+        self.chakra_host_trace_path = self.trace_dir / "host"
+        self.chakra_device_trace_path = self.trace_dir / "device"
+
+        self.chakra_host_trace_path.mkdir(parents=True, exist_ok=True)
+        self.chakra_device_trace_path.mkdir(parents=True, exist_ok=True)
+
+        self.trace_observer = torch.profiler.ExecutionTraceObserver()
+
+        def trace_handler(prof):
```
Comment (titu1994) [inline nemo/lightning/pytorch/callbacks/chakra.py:71]: Inner function cannot be serialized, is it necessary? Maybe move it out.

**Review by titu1994 — COMMENTED** [review]
Looks reasonable

File: nemo/lightning/pytorch/callbacks/chakra.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        warmup_steps: int = 0,
+        active_steps: int = 1,
+        trace_dir: str = None,
+    ):
+        if not isinstance(start_step, int) or not isinstance(end_step, int):
+            raise TypeError(f"start_step and end_step must be integers. Got {type(start_step)}, {type(end_step)}")
+        if end_step < start_step:
+            raise ValueError("end_step must be greater than or equal to start_step.")
+
+        if not trace_dir or not os.path.isdir(trace_dir):
```
Comment (maanug-nv) [inline nemo/lightning/pytorch/callbacks/chakra.py:54]: If `trace_dir` is required, should it have no default?

**Comment by maanug-nv** [comment]
Adding a short test would be good as well, to ensure all files that are expected to be created do exist. Fine with a separate PR for that

**Comment by TaekyungHeo** [comment]
Thanks for the reminder — I'll get back to this, but I have some higher-priority items at the moment.

File: nemo/lightning/pytorch/callbacks/chakra.py (hunk 1 above)
Comment (TaekyungHeo) [reply nemo/lightning/pytorch/callbacks/chakra.py]: @titu1994 , I renamed it to `PytorchProfilerCallback`

File: nemo/lightning/pytorch/callbacks/chakra.py (hunk 3 above)
Comment (TaekyungHeo) [reply nemo/lightning/pytorch/callbacks/chakra.py]: Updated the logic as follows: if trace_dir is None, a directory will be created under the current working directory.
```
    def __init__(
        self,
        start_step: int,
        end_step: int,
        warmup_steps: int = 0,
        active_steps: int = 1,
        trace_dir: str = None,
    ):
        if trace_dir is None:
            trace_dir = os.path.join(os.getcwd(), 'traces')
            os.makedirs(trace_dir, exist_ok=True)
```

File: nemo/lightning/pytorch/callbacks/chakra.py (hunk 2 above)
Comment (TaekyungHeo) [reply nemo/lightning/pytorch/callbacks/chakra.py]: @titu1994 Updated.

**Review by titu1994 — COMMENTED** [review]
Looks good, lets wait for @maanug-nv  for final signoff

**Review by maanug-nv — DISMISSED** [review]
LGTM, will take testing discussion offline

**Comment by maanug-nv** [comment]
merged in main again to re-trigger readthedocs build

**Review by maanug-nv — APPROVED** [review]
re-approving

**Comment by chtruong814** [comment]
We'll fast merge this since it passed the unit tests and this is new code that shouldn't be used anywhere right now. We'll keep an eye out if that previous test failure resurfaces.
